opm/simulators/wells/WellAndGroupStates.hpp

```diff
@@ -0,0 +1,564 @@
+/// communicated as needed).
+template <int NumActivePhases>
+class WellAndGroupStates
+{
+public:
+
+    void init(const std::vector<double>& cell_pressures,
+              const Schedule& schedule,
+              const std::vector<Well>& wells,
+              const int report_step,
+              const WellAndGroupStates* prev_state,
+              const PhaseUsage& phase_usage,
+              const std::vector<std::vector<PerforationData>>& well_perf_data,
+              const SummaryState& summary_state)
+    {
+        assert(wells.size() == well_perf_data.size());
+        well_perf_data_ = well_perf_data;
+
+        // Initialize all wells.
+        const int nw = wells.size();
+        well_states_.resize(nw);
+        well_names_.resize(nw);
+        for (int w = 0; w < nw; ++w) {
+            well_names_[w] = wells[w].name();
+            initSingleWell(cell_pressures, schedule, wells[w], report_step, phase_usage, well_perf_data_[w], summary_state, well_states_[w]);
+        }
+
+        // For wells present in a previous state, override initialization where appropriate.
+        if (prev_state) {
+            for (int w = 0; w < nw; ++w) {
+                well_names_[w] = wells[w].name();
+                auto it = std::find(prev_state->well_names_.begin(), prev_state->well_names_.end(), well_names_[w]);
+                if (it != prev_state->well_names_.end()) {
+                    const int w_in_prev = it - prev_state->well_names_.end();
+                    initSingleWellFromPrevious(prev_state->well_states_[w_in_prev], well_states_[w]);
+                }
+            }
+        }
+
+        // Group states remain default-initialized.
+    }
+
+    data::Wells report(const PhaseUsage& phase_usage, const int* globalCellIdxMap) const
+    {
+        data::Wells dwells;
+        const int nw = well_states_.size();
+        for (int w = 0; w < nw; ++w) {
+            if (well_states_[w].status == Well::Status::SHUT) {
+                continue;
+            }
+            auto& dwell = dwells[well_names_[w]];
+            reportSingleWell(phase_usage, globalCellIdxMap, well_perf_data_[w], well_states_[w], dwell);
+        }
+        return dwells;
+    }
+
+    using WellState = SingleWellState<NumActivePhases>;
+    using GroupState = SingleGroupState<NumActivePhases>;
+
+    std::vector<WellState>& wellStates()
+    {
+        return well_states_;
+    }
+    const std::vector<WellState>& wellStates() const
+    {
+        return well_states_;
+    }
+
+    std::map<std::string, GroupState>& groupStates()
+    {
+        return group_states_;
+    }
+    const std::map<std::string, GroupState>& groupStates() const
+    {
+        return group_states_;
+    }
+
+
+private:
+
+    // -----------  Data members  -----------
+
+    std::vector<WellState> well_states_;
+    std::vector<std::string> well_names_;
+    std::map<std::string, GroupState> group_states_;
+
+    std::vector<std::vector<PerforationData>> well_perf_data_;
```

Comment:
The `PerforationData` could be part of `WellState`?

Reply:
I have tried to put only dynamic data into SingleWellState, keeping the static data (PerforationData, should be called StaticConnectionData perhaps) separate.

Reply:
OK - that makes sense, but if the connection data is static they could just come from opm-common?

Reply:
Not the cell index at least (if you recall, we discussed a while back what was needed in addition to the opm-common provided static info, and the cell index is the most important).

Reply:
Yes - I briefly remember that; then maybe in the future we could refactor here - and make some alterations to the opm-common datastructure - but not now I guess.